### PR TITLE
Add a port number guard and use that in the checks instead of range

### PR DESCRIPTION
**Pull request by ryanwinchester**
I talked about it at the end of #145 although it's a micro-optimization.

I didn't know where to put it, I could put it as a private guard in the headers module? Or somewhere else?

### CODE REVIEW

**Comment by mtrudel** [comment]
This is an extremely hot path (one of the hottest), so though it's a micro optimization it's an important one!

File: lib/bandit/ports.ex [hunk 1]
```diff
@@ -0,0 +1,21 @@
+defmodule Bandit.Ports do
+  @moduledoc """
+  Bandit helper functions and guards for dealing with port numbers.
+  """
```
Comment (mtrudel) [inline lib/bandit/ports.ex:4]: Internal libs like this use `@moduledoc false` to keep the docs focused on user-visible parts of Bandit.

File: lib/bandit/ports.ex [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+  ## Examples
+
+      iex> is_port_number(443)
+      true
+
+      iex> is_port_number(-1)
+      false
+
+  """
```
Comment (mtrudel) [inline lib/bandit/ports.ex:19]: ibid.

File: lib/bandit/headers.ex [hunk 3]
```diff
@@ -2,6 +2,8 @@ defmodule Bandit.Headers do
   @moduledoc false
   # Conveniences for dealing with headers
 
+  import Bandit.Ports
+
```
Comment (mtrudel) [inline lib/bandit/headers.ex:6]: The implementation of this is pretty simple; I think this is a rare case of `defguardp`

File: test/bandit/http1/request_test.exs [hunk 4]
```diff
@@ -108,6 +108,64 @@ defmodule HTTP1RequestTest do
       assert Jason.decode!(body)["port"] == 1234
     end
 
+    test "derives port from host header for all valid ports", context do
+      client = SimpleHTTP1Client.tcp_client(context)
+
+      for port <- 0..65_535 do
```
Comment (mtrudel) [inline test/bandit/http1/request_test.exs:114]: This is a lot of requests! We could probably stand to do a subset of this (either random, some set of pre-selected ones, or using a `//step` on the range

Comment (ryanwinchester) [reply test/bandit/http1/request_test.exs]: I kind of thought that's what you intended when you wanted my other tests ported to these modules, but I pushed a more sane version.

File: lib/bandit/headers.ex [hunk 5]
```diff
@@ -2,6 +2,8 @@ defmodule Bandit.Headers do
   @moduledoc false
   # Conveniences for dealing with headers
 
+  defguardp is_port_number(port) when Bitwise.band(port, 0xFFFF) === port
+
```
Comment (mtrudel) [inline lib/bandit/headers.ex:6]: Nit, but if this is private oughtn't it be at the bottom of the file?

Comment (ryanwinchester) [reply lib/bandit/headers.ex:6]: guards have to be defined above where you use them

Comment (ryanwinchester) [reply lib/bandit/headers.ex:6]: ^

**Comment by ryanwinchester** [comment]
I've been playing around with another micro-optimization for this module. Another PR and benchmarks incoming when this is merged.

**Pull request merged**